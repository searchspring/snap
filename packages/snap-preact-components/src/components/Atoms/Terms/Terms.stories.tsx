/** @jsx h */
import { h } from 'preact';

import { ArgsTable, PRIMARY_STORY } from '@storybook/addon-docs/blocks';

import { componentArgs } from '../../../utilities';
import { Terms, TermsProps } from './Terms';
import Readme from './readme.md';
import { AutocompleteController } from '@searchspring/snap-controller';
import { Snapify } from '../../../utilities/snapify';

export default {
	title: `Atoms/Terms`,
	component: Terms,
	parameters: {
		docs: {
			page: () => (
				<div>
					<Readme />
					<ArgsTable story={PRIMARY_STORY} />
				</div>
			),
		},
	},
	decorators: [
		(Story: any) => (
			<div
				style={{
					maxWidth: '900px',
					position: 'relative',
				}}
			>
				<input
					type="text"
					id="searchInput"
					placeholder="try me!"
					autoComplete="off"
					style={{
						width: '100%',
						padding: '10px',
						boxSizing: 'border-box',
						border: '1px solid #3a23ad',
					}}
				/>
				<Story />
			</div>
		),
	],
	argTypes: {
		controller: {
			description: 'Autocomplete controller reference',
			type: { required: false },
			table: {
				type: {
					summary: 'Autocomplete controller object',
				},
			},
			control: { type: 'none' },
		},
		terms: {
			description: 'Autocomplete Term store reference',
			type: { required: false },
			table: {
				type: {
					summary: 'Autocomplete Term store object',
				},
			},
			control: { type: 'none' },
		},
		title: {
			description: 'terms title',
			defaultValue: '',
			type: { required: false },
			table: {
				type: {
					summary: 'string',
				},
			},
			control: { type: 'text' },
		},
		limit: {
			description: 'adjust the number of terms to show',
			defaultValue: '',
			type: { required: false },
			table: {
				type: {
					summary: 'number',
				},
			},
			control: { type: 'number' },
		},
		previewOnHover: {
			defaultValue: false,
			description: 'Invoke term preview upon focus',
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: false },
			},
			control: { type: 'boolean' },
		},
		emIfy: {
			defaultValue: false,
			description: 'highlight the query matched section of the term',
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: false },
			},
			control: { type: 'boolean' },
		},
		onTermClick: {
			description: 'Custom onClick event handler for Terms',
			table: {
				type: {
					summary: 'function',
				},
			},
			action: 'onTermClick',
		},
		...componentArgs,
	},
};

const snapInstance = Snapify.autocomplete({
	id: 'Autocomplete',
	selector: '#searchInput',
	globals: {
		siteId: '8uyt2m',
	},
	settings: {
		trending: {
			limit: 5,
		},
	},
});

export const Default = (args: TermsProps, { loaded: { controller } }: { loaded: { controller: AutocompleteController } }) => {
	setTimeout(() => {
		controller.bind();
	});
	return <Terms {...args} controller={controller} />;
};

Default.loaders = [
	async () => ({
		controller: await snapInstance,
	}),
];
