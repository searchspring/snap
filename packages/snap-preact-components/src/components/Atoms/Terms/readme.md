# Terms

Renders a list of terms for autocomplete. 

## Usage

### Controller

The `controller` prop specifies a reference to the autocomplete controller.

```jsx
<Terms controller={controller} />
```

### Terms

The `terms` prop specifies a reference to an autocomplete term store. 

```jsx
<Terms controller={controller} terms={terms}/>
```

### title

The `title` prop is will display the given text above the terms area. The default value is blank. 

```jsx
<Terms controller={controller} terms={terms} title={'Suggestions'}/>
```

### limit

The `limit` prop is used for number of terms to show. 

```jsx
<Terms controller={controller} terms={terms} limit={4}/>
```

### previewOnHover

The `previewOnFocus` prop will invoke the `term.preview()` method when the value has been hovered over.

```jsx
<Terms controller={controller} terms={terms} previewOnHover={true}/>
```

### emify

The `emify` will automatically wrap non-matching characters within the term with an `<em>` element for different styling purposes. 

```jsx
<Terms controller={controller} terms={terms} emify={true}/>
```

### onTermClick

The `onTermClick` prop allows for a custom callback function for when a term value is clicked.

```jsx
<Terms controller={controller} terms={terms} onTermClick={(e)=>{console.log(e)}}/>
```
