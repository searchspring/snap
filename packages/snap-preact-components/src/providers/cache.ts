export { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';

export const cache = createCache({
	key: 'ss',
	prepend: true,
});
