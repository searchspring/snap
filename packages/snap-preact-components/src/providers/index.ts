export * from './store';
export * from './theme';
