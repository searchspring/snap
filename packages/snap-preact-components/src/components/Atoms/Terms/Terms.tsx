/** @jsx jsx */
import { Fragment, h } from 'preact';

import { observer } from 'mobx-react-lite';
import { jsx, css } from '@emotion/react';
import classnames from 'classnames';

import type { AutocompleteController } from '@searchspring/snap-controller';
import type { AutocompleteTermStore } from '@searchspring/snap-store-mobx/dist/cjs/Autocomplete/Stores';
import { ComponentProps, StylingCSS } from '../../../types';
import { Theme, useTheme, CacheProvider } from '../../../providers';
import { createHoverProps } from '../../../toolbox';
import { mergeProps } from '../../../utilities';

const CSS = {
	Terms: () => css({}),
};

export const Terms = observer((properties: TermsProps): JSX.Element => {
	const globalTheme: Theme = useTheme();
	const defaultProps: Partial<TermsProps> = {};

	const props = mergeProps('terms', globalTheme, defaultProps, properties);

	const { terms, title, onTermClick, limit, previewOnHover, emIfy, disableStyles, style, controller } = props;
	const currentInput = controller?.store?.state?.input;

	const styling: { css?: StylingCSS } = {};
	if (!disableStyles) {
		styling.css = [CSS.Terms(), style];
	} else if (style) {
		styling.css = [style];
	}

	const emIfyTerm = (term: string, search: string) => {
		if (term && search) {
			const match = term.match(escapeRegExp(search));
			if (search && term && match && typeof match.index == 'number') {
				const beforeMatch = term.slice(0, match.index);
				const afterMatch = term.slice(match.index + search.length, term.length);
				return (
					<>
						{beforeMatch ? <em>{beforeMatch}</em> : ''}
						{search}
						{afterMatch ? <em>{afterMatch}</em> : ''}
					</>
				);
			}
		}

		return (
			<Fragment>
				<em>{term}</em>
			</Fragment>
		);
	};

	const termClickEvent = (e: React.MouseEvent<Element, MouseEvent>) => {
		onTermClick && onTermClick(e);

		// remove focus from input (close the autocomplete)
		controller?.setFocused && controller?.setFocused();
	};

	const escapeRegExp = (string: string): string => {
		return string?.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
	};

	const termsToShow = limit ? terms?.slice(0, limit) : terms;

	return termsToShow?.length ? (
		<CacheProvider>
			<div className="ss__autocomplete__terms">
				{title ? (
					<div className="ss__autocomplete__title">
						<h5>{title}</h5>
					</div>
				) : null}
				<div className="ss__autocomplete__terms__options" role={'list'} aria-label={title}>
					{termsToShow?.map((term, idx) => (
						<div
							className={classnames('ss__autocomplete__terms__option', {
								'ss__autocomplete__terms__option--active': term.active,
							})}
						>
							<a
								onClick={(e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => termClickEvent(e)}
								href={term.url.href}
								{...(previewOnHover ? createHoverProps(term.preview) : {})}
								role="link"
								aria-label={`item ${idx + 1} of ${history.length}, ${term.value}`}
							>
								{emIfy ? emIfyTerm(term.value, currentInput || '') : term.value}
							</a>
						</div>
					))}
				</div>
			</div>
		</CacheProvider>
	) : (
		<Fragment></Fragment>
	);
});

export interface TermsProps extends ComponentProps {
	controller?: AutocompleteController;
	terms?: AutocompleteTermStore;
	title?: string;
	limit?: number;
	onTermClick?: (e: React.MouseEvent<Element, MouseEvent>) => void;
	previewOnHover?: boolean;
	emIfy?: boolean;
}
