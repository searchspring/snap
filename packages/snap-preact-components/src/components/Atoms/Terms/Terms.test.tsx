import { h } from 'preact';

import { v4 as uuidv4 } from 'uuid';
import { render } from '@testing-library/preact';
import { ThemeProvider } from '../../../providers';

import { Terms } from './Terms';
import { AutocompleteController, AutocompleteControllerConfig } from '@searchspring/snap-controller';
import { createAutocompleteController } from '@searchspring/snap-preact';
import { MockClient } from '@searchspring/snap-shared';
import userEvent from '@testing-library/user-event';

describe('Terms Component', () => {
	const globals = { siteId: '8uyt2m' };
	let controllerConfigId = uuidv4().split('-').join('');
	let acConfig: AutocompleteControllerConfig = {
		id: controllerConfigId,
		selector: 'input.searchspring-ac',
		settings: {
			trending: {
				limit: 5,
			},
			history: {
				limit: 5,
			},
		},
	};
	const clientConfig = {
		globals: {
			siteId: '8uyt2m',
		},
	};

	const mockClient = new MockClient(globals, {});
	mockClient.mockData.updateConfig({ meta: 'ac.meta' });

	let mockTerms = [
		{
			active: false,
			preview: () => {},
			value: 'dress',
			url: {
				href: 'www.dress.com',
			},
		},
		{
			active: false,
			preview: () => {},
			value: 'drss',
			url: {
				href: 'www.drss.com',
			},
		},
		{
			active: false,
			preview: () => {},
			value: 'dres',
			url: {
				href: 'www.dres.com',
			},
		},
		{
			active: false,
			preview: () => {},
			value: 'dss',
			url: {
				href: 'www.dss.com',
			},
		},
		{
			active: false,
			preview: () => {},
			value: 'ress',
			url: {
				href: 'www.ress.com',
			},
		},
	];

	let controller: AutocompleteController;
	beforeEach(async () => {
		controller = createAutocompleteController({ client: clientConfig, controller: acConfig }, { client: mockClient });
		await controller.bind();
	});

	it('renders', async () => {
		// @ts-ignore - typing on terms
		const rendered = render(<Terms controller={controller} terms={mockTerms} />);
		const terms = rendered.container.querySelector('.ss__terms');
		expect(terms).toBeInTheDocument();
	});

	it('renders with a title', async () => {
		const title = 'Suggestions';
		// @ts-ignore - typing on terms
		const rendered = render(<Terms controller={controller} terms={mockTerms} title={title} />);
		const terms = rendered.container.querySelector('.ss__terms');
		const titleElem = rendered.container.querySelector('.ss__terms__title');
		expect(terms).toBeInTheDocument();
		expect(titleElem).toBeInTheDocument();
		expect(titleElem).toHaveTextContent(title);
	});

	it('respects the limit', async () => {
		const limit = 3;
		// @ts-ignore - typing on terms
		const rendered = render(<Terms controller={controller} terms={mockTerms} limit={limit} />);
		const termOptions = rendered.container.querySelectorAll('.ss__terms__option');
		expect(termOptions).toHaveLength(limit);
	});

	it('Can set custom onClick', async () => {
		const onClick = jest.fn();
		// @ts-ignore - typing on terms
		const rendered = render(<Terms controller={controller} terms={mockTerms} onTermClick={onClick} />);
		const termOptions = rendered.container.querySelector('.ss__terms__option a');

		userEvent.click(termOptions!);

		expect(onClick).toHaveBeenCalled();
	});

	it('Can use emify', async () => {
		document.body.innerHTML = '<div>' + '  <input type="text" class="searchspring-ac">' + '<div id="target"></div></div>';
		mockClient.mockData.updateConfig({ autocomplete: 'default' });

		const input = document.querySelector('.searchspring-ac');
		expect(input).toBeInTheDocument();

		await controller.bind();

		userEvent.type(input!, 'dre');

		// @ts-ignore - typing on terms
		const rendered = render(<Terms controller={controller} terms={mockTerms} emIfy={true} />);
		const termOptions = rendered.container.querySelectorAll('.ss__terms__option a')[0];

		expect(termOptions?.innerHTML).toBe('dre<em>ss</em>');
	});

	it('renders with classname', () => {
		const className = 'classy';

		// @ts-ignore - typing on terms
		const rendered = render(<Terms controller={controller} terms={mockTerms} className={className} />);
		const terms = rendered.container.querySelector('.ss__terms');

		expect(terms).toHaveClass(className);
	});

	it('disables styles', () => {
		// @ts-ignore - typing on terms
		const rendered = render(<Terms controller={controller} terms={mockTerms} disableStyles />);
		const terms = rendered.container.querySelector('.ss__terms');

		expect(terms?.classList).toHaveLength(1);
	});

	it('is themeable with ThemeProvider', () => {
		const globalTheme = {
			components: {
				terms: {
					disableStyles: true,
				},
			},
		};
		const rendered = render(
			<ThemeProvider theme={globalTheme}>
				{/* @ts-ignore - typing on terms */}
				<Terms controller={controller} terms={mockTerms} />
			</ThemeProvider>
		);
		const terms = rendered.container.querySelector('.ss__terms');
		expect(terms).toBeInTheDocument();
		expect(terms?.classList.length).toBe(1);
	});

	it('is themeable with theme prop', () => {
		const propTheme = {
			components: {
				terms: {
					disableStyles: true,
				},
			},
		};
		const rendered = render(
			// @ts-ignore - typing on terms
			<Terms controller={controller} terms={mockTerms} theme={propTheme} />
		);
		const terms = rendered.container.querySelector('.ss__terms');
		expect(terms).toBeInTheDocument();
		expect(terms?.classList.length).toBe(1);
	});

	it('is theme prop overrides ThemeProvider', () => {
		const globalTheme = {
			components: {
				terms: {
					disableStyles: false,
				},
			},
		};
		const propTheme = {
			components: {
				terms: {
					disableStyles: true,
				},
			},
		};
		const rendered = render(
			<ThemeProvider theme={globalTheme}>
				{/* @ts-ignore - typing on terms  */}
				<Terms controller={controller} terms={mockTerms} theme={propTheme} />
			</ThemeProvider>
		);

		const terms = rendered.container.querySelector('.ss__terms');
		expect(terms).toBeInTheDocument();
		expect(terms?.classList.length).toBe(1);
	});
});
