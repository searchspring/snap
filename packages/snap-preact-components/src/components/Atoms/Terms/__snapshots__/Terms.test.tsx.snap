// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terms Component Can set custom onClick 1`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component Can set custom onClick 2`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component Can use emify 1`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dre
          <em>
            ss
          </em>
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          <em>
            drss
          </em>
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dre
          <em>
            ees
          </em>
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dre
          <em>
            s
          </em>
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          <em>
            dss
          </em>
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          <em>
            ress
          </em>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component disables styles 1`] = `
<DocumentFragment>
  <div
    class="ss__terms"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component doesnt render with no controller or terms 1`] = `<DocumentFragment />`;

exports[`Terms Component is theme prop overrides ThemeProvider 1`] = `
<DocumentFragment>
  <div
    class="ss__terms"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component is themeable with ThemeProvider 1`] = `
<DocumentFragment>
  <div
    class="ss__terms"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component is themeable with theme prop 1`] = `
<DocumentFragment>
  <div
    class="ss__terms"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component passed in terms take priority over controller terms 1`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component previewonhover 1`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component previewonhover 2`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component renders 1`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component renders with a title 1`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__title"
    >
      <h5>
        Suggestions
      </h5>
    </div>
    <div
      aria-label="Suggestions"
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component renders with classname 1`] = `
<DocumentFragment>
  <div
    class="ss__terms classy emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component renders with terms just in controller 1`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 6, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 6, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 6, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 4 of 6, dres"
          href="www.dres.com"
          role="link"
        >
          dres
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 5 of 6, dss"
          href="www.dss.com"
          role="link"
        >
          dss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 6 of 6, ress"
          href="www.ress.com"
          role="link"
        >
          ress
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Terms Component respects the limit 1`] = `
<DocumentFragment>
  <div
    class="ss__terms emotion-0"
  >
    <div
      class="ss__terms__options"
      role="list"
    >
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 1 of 3, dress"
          href="www.dress.com"
          role="link"
        >
          dress
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 2 of 3, drss"
          href="www.drss.com"
          role="link"
        >
          drss
        </a>
      </div>
      <div
        class="ss__terms__option"
      >
        <a
          aria-label="item 3 of 3, dreees"
          href="www.dreees.com"
          role="link"
        >
          dreees
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
